### ISSUE TITLE
[1.13.0]: docker network rm: network has active endpoints

### ISSUE THREAD

**Issue by hmdhasani** [moby/moby#29631]
When I want to remove a container some times I see this error:

```
Error response from daemon: Unable to remove filesystem for 5e32b44ce67eee61b954138f4bb4a86e72fe1ac335fcd83065902f44441b327a: remove /var/lib/docker/containers/5e32b44ce67eee61b954138f4bb4a86e72fe1ac335fcd83065902f44441b327a/shm: device or resource busy
```
but it seems that the container is removed as its not in `docker ps -a`

but when I want to remove the network it gave me:

    Error response from daemon: network net_c36-150-0 has active endpoints

```
docker network inspect net_c36-150-0
[
    {
        "Name": "net_c36-150-0",
        "Id": "5f7db9d5db60178cbfe02a2a14edae8d6062b3fbf883f5266d3589a882816e77",
        "Created": "2016-12-20T17:24:38.377424861+03:30",
        "Scope": "local",
        "Driver": "bridge",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": {},
            "Config": [
                {
                    "Subnet": "21.1.1.0/16"
                }
            ]
        },
        "Internal": false,
        "Attachable": false,
        "Containers": {
            "10954a48bd4e194417f0a0e405d6a02cbee837c9d2200265845935b7f5b31e21": {
                "Name": "cassandra-1_c36-150-0",
                "EndpointID": "a6df2d18569c1794fc9e2727b293757375c94034ebf630394326fbd471494652",
                "MacAddress": "02:42:15:01:00:0f",
                "IPv4Address": "21.1.0.15/16",
                "IPv6Address": ""
            }
        },
        "Options": {},
        "Labels": {}
    }
]
```

**Comment by hmdhasani** [comment]
docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 89
Server Version: 1.13.0-rc4
Storage Driver: overlay
 Backing Filesystem: extfs
 Supports d_type: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 03e5862ec0d8d3b3f750e19fca3ee367e13c090e
runc version: 51371867a01c467f08af739783b8beafc154c4d7
init version: 949e6fa
Security Options:
 apparmor
Kernel Version: 3.19.0-78-generic
Operating System: Ubuntu 14.04.5 LTS
OSType: linux
Architecture: x86_64
CPUs: 12
Total Memory: 19.61 GiB
Name: s36
ID: NDW6:JIM5:QO7M:E7BH:SCJP:DSUJ:HORQ:XB6B:N7LF:U57V:HP4W:RKI3
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

**Comment by hmdhasani** [comment]
docker version
Client:
 Version:      1.13.0-rc4
 API version:  1.25
 Go version:   go1.7.3
 Git commit:   88862e7
 Built:        Fri Dec 16 22:59:15 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.13.0-rc4
 API version:  1.25 (minimum version 1.12)
 Go version:   go1.7.3
 Git commit:   88862e7
 Built:        Fri Dec 16 22:59:15 2016
 OS/Arch:      linux/amd64
 Experimental: false

**Comment by hmdhasani** [comment]
Update: restarting docker service solves the problem

**Comment by tonistiigi** [comment]
@dongluochen I think you looked into something similar

**Comment by dongluochen** [comment]
@hmdhasani There are multiple issues with "Unable to remove filesystem". Can you take a look if this is covered by https://docs.docker.com/engine/admin/troubleshooting_volume_errors/, #22260, or #25718?

**Comment by tonistiigi** [comment]
@dongluochen I actually meant the `network net_c36-150-0 has active endpoints` error that is somehow related to swarm services.

**Comment by hmdhasani** [comment]
@dongluochen it doesn't seem to be  https://docs.docker.com/engine/admin/troubleshooting_volume_errors/ or , #22260, or #25718 because in those cases the container could not be removed but in my case the container is removed at least from "docker ps -a" result. but I think there are still some side effects on network.
The reason why I first thought it was a 1.13.0-rc4 bug was that it only occurs on one of our test servers that has this version. but I also realized that the server has less RAM than others and now I have doubt whether it's a 1.13.0-rc4 bug or a general one. I am trying to reproduce the error, but it's a very rare one.

**Comment by tonistiigi** [comment]
@mlaventure Failure to delete the `shm` should point to container process still being alive. Even if only mounts were alive it shouldn't cause this. So it may be a containerd sync issue.

**Comment by mlaventure** [comment]
@hmdhasani do you have daemon debug log around the time that issue occurred?

**Comment by hmdhasani** [comment]
@mlaventure currently no but I will get that

**Comment by hmdhasani** [comment]
I installed docker 1.12.5 on the same server that had 1.13.0-rc4 and do the same procedure for many times and I never got the error, so currently I am almost sure that this is a 1.13.0-rc4 bug.
I will set the debug option to get more info.

**Comment by hmdhasani** [comment]
I reinstalled 1.13.0-rc4 and got the error back:


This is stdout:

docker rm -f -v cassandra-1_c36-122-0
Error response from daemon: Unable to remove filesystem for 0faf00e710365f2f3b3fcc6275e2f24dbf4ec5fce69a3a0dc571c6e414424516: remove /var/lib/docker/containers/0faf00e710365f2f3b3fcc6275e2f24dbf4ec5fce69a3a0dc571c6e414424516/shm: device or resource busy



And This is Debug log:

time="2017-01-07T10:57:02.457407250+03:30" level=debug msg="Calling GET /v1.24/containers/89b7924dfdbc/json"
time="2017-01-07T10:57:02.480756222+03:30" level=debug msg="Calling GET /v1.24/containers/44f7b609d9c2/json"
time="2017-01-07T10:57:02.509474858+03:30" level=debug msg="Calling DELETE /v1.24/containers/cassandra-0_c36-122-0?force=1&v=1"
time="2017-01-07T10:57:02.646991895+03:30" level=debug msg="Removing volume reference: driver local, name 7c897af97da79f1ceb5390a80a2d8f22c8ae6ec29970a07e9f465baf9d6d697a"
time="2017-01-07T10:57:02.989828755+03:30" level=debug msg="Calling GET /v1.24/containers/json?all=1"
time="2017-01-07T10:57:03.015207896+03:30" level=debug msg="Calling GET /v1.24/containers/3da8af17ef9d/json"
time="2017-01-07T10:57:03.020127898+03:30" level=debug msg="Read from DNS server failed, read udp 21.1.0.20:39640->194.225.0.14:53: i/o timeout"
time="2017-01-07T10:57:03.020127224+03:30" level=debug msg="Read from DNS server failed, read udp 21.1.0.20:46317->194.225.0.14:53: i/o timeout"
time="2017-01-07T10:57:03.038117498+03:30" level=debug msg="Calling GET /v1.24/containers/08d21d93c81e/json"
time="2017-01-07T10:57:03.059291841+03:30" level=debug msg="Calling GET /v1.24/containers/ea264394e521/json"
time="2017-01-07T10:57:03.090701680+03:30" level=debug msg="Calling GET /v1.24/containers/0faf00e71036/json"
time="2017-01-07T10:57:03.113944623+03:30" level=debug msg="Calling GET /v1.24/containers/e6e8bd5e15bb/json"
time="2017-01-07T10:57:03.141951588+03:30" level=debug msg="Name To resolve: apache-center-rajae."
time="2017-01-07T10:57:03.142110491+03:30" level=debug msg="Lookup for apache-center-rajae.: IP [21.1.0.7]"
time="2017-01-07T10:57:03.141920262+03:30" level=debug msg="Name To resolve: apache-center-rajae."
time="2017-01-07T10:57:03.142193988+03:30" level=debug msg="Lookup name apache-center-rajae. present without IPv6 address"
time="2017-01-07T10:57:03.142694902+03:30" level=debug msg="Calling GET /v1.24/containers/393899f5b419/json"
time="2017-01-07T10:57:03.164022842+03:30" level=debug msg="Calling GET /v1.24/containers/231ecc95475b/json"
time="2017-01-07T10:57:03.184762050+03:30" level=debug msg="Calling GET /v1.24/containers/7c67b792f5e7/json"
time="2017-01-07T10:57:03.205098912+03:30" level=debug msg="Calling GET /v1.24/containers/9e44f2e73fe7/json"
time="2017-01-07T10:57:03.226039675+03:30" level=debug msg="Calling GET /v1.24/containers/ff7e3f4572d6/json"
time="2017-01-07T10:57:03.245674576+03:30" level=debug msg="Calling GET /v1.24/containers/3a788ac6d1da/json"
time="2017-01-07T10:57:03.278857746+03:30" level=debug msg="Calling GET /v1.24/containers/bfd2313713d4/json"
time="2017-01-07T10:57:03.307149760+03:30" level=debug msg="Calling GET /v1.24/containers/b5409088e66f/json"
time="2017-01-07T10:57:03.329743668+03:30" level=debug msg="Calling GET /v1.24/containers/e2e5adb877a9/json"
time="2017-01-07T10:57:03.351918760+03:30" level=debug msg="Calling GET /v1.24/containers/415cb96a649d/json"
time="2017-01-07T10:57:03.364826086+03:30" level=debug msg="Name To resolve: mysql-rajae.ut.ac.ir."
time="2017-01-07T10:57:03.364964316+03:30" level=debug msg="Name To resolve: mysql-rajae.ut.ac.ir."
time="2017-01-07T10:57:03.365183379+03:30" level=debug msg="Query mysql-rajae.ut.ac.ir.[28] from 21.1.0.21:34414, forwarding to udp:192.168.20.14"
time="2017-01-07T10:57:03.365183431+03:30" level=debug msg="Query mysql-rajae.ut.ac.ir.[1] from 21.1.0.21:33116, forwarding to udp:192.168.20.14"
time="2017-01-07T10:57:03.380366972+03:30" level=debug msg="Calling GET /v1.24/containers/ad752abc7b95/json"
time="2017-01-07T10:57:03.400803066+03:30" level=debug msg="Calling GET /v1.24/containers/b62a78951a6c/json"
time="2017-01-07T10:57:03.423255936+03:30" level=debug msg="Calling GET /v1.24/containers/d5f4b9d46295/json"
time="2017-01-07T10:57:03.469087170+03:30" level=debug msg="Calling GET /v1.24/containers/97f8b2993348/json"
time="2017-01-07T10:57:03.491450475+03:30" level=debug msg="Calling GET /v1.24/containers/e09cf94ff4e6/json"
time="2017-01-07T10:57:03.536606710+03:30" level=debug msg="Calling GET /v1.24/containers/bc24aba63bdd/json"
time="2017-01-07T10:57:03.564514417+03:30" level=debug msg="Calling GET /v1.24/containers/f73542dd447c/json"
time="2017-01-07T10:57:03.585062547+03:30" level=debug msg="Calling GET /v1.24/containers/a9691f7a4271/json"
time="2017-01-07T10:57:03.607540618+03:30" level=debug msg="Calling GET /v1.24/containers/0b41dde22903/json"
time="2017-01-07T10:57:03.636442691+03:30" level=debug msg="Calling GET /v1.24/containers/d66b66aa8f13/json"
time="2017-01-07T10:57:03.657547399+03:30" level=debug msg="Calling GET /v1.24/containers/6f601e0dcd5c/json"
time="2017-01-07T10:57:03.679584285+03:30" level=debug msg="Calling GET /v1.24/containers/adff9020d0d8/json"
time="2017-01-07T10:57:03.700132268+03:30" level=debug msg="Calling GET /v1.24/containers/89b7924dfdbc/json"
time="2017-01-07T10:57:03.721183865+03:30" level=debug msg="Calling GET /v1.24/containers/44f7b609d9c2/json"
time="2017-01-07T10:57:03.742016737+03:30" level=debug msg="Calling DELETE /v1.24/containers/cassandra-1_c36-122-0?force=1&v=1"
time="2017-01-07T10:57:03.742167827+03:30" level=debug msg="Sending kill signal 9 to container 0faf00e710365f2f3b3fcc6275e2f24dbf4ec5fce69a3a0dc571c6e414424516"
time="2017-01-07T10:57:03.743404775+03:30" level=warning msg="container kill failed because of 'container not found' or 'no such process': Cannot kill container 0faf00e710365f2f3b3fcc6275e2f24dbf4ec5fce69a3a0dc571c6e414424516: rpc error: code = 2 desc = containerd: container not found"
time="2017-01-07T10:57:04.023271766+03:30" level=debug msg="Read from DNS server failed, read udp 21.1.0.20:57139->192.168.20.14:53: i/o timeout"
time="2017-01-07T10:57:04.023315726+03:30" level=debug msg="Read from DNS server failed, read udp 21.1.0.20:36875->192.168.20.14:53: i/o timeout"

**Comment by hmdhasani** [comment]
bad news: It seems that this bug exists in 1.13.0

**Comment by hmdhasani** [comment]
Update: This bug occurs only after some hours of running a container. for example as I measure there is   about 10% chance to have a container like this after 24 hours of running it.

**Comment by hmdhasani** [comment]
I think this bug must get some priority!
It made us revert to 1.12

**Comment by tonistiigi** [comment]
@mlaventure Has to be a containerd problem afaics. `rpc error: code = 2 desc = containerd: container not found` but for the rm it is still alive. Would `ps aux` help when this happens or a stack trace from containerd?

@hmdhasani Is this happening on specific containers? Could help if you could send a longer debug logs so we see what is happening with the processes before this happens. For example in the excerpt you posted there are no event logs from containerd.

**Comment by mlaventure** [comment]
@hmdhasani are you using live restore?

Do you have more logs available? It looks like the container must have "exited" already, in which case the events should have been seen by containerd.

Could you share the `/run/docker/libcontainerd/containerd/events.log` file if you still have that instance running?

the output of `ps aux` could help to check if the container process is still alive, but we'd need the expected pid from `docker inspect` to ascertain that.

**Comment by tonistiigi** [comment]
We managed to reproduce this issue in combination with #30469 . The issue here appears from improper handling of containerd errors but the specific error that could make it a `v1.13` regression could be #30469. @hmdhasani We will let you know when fix for #30469 is vendored into Docker, it would be good if you could verify if this solves this problem as well for you.

**Comment by hmdhasani** [comment]
This is a complete log:
[docker-log.zip](https://github.com/docker/docker/files/740032/docker-log.zip)

**Comment by tonistiigi** [comment]
@mlaventure 
```
Jan 25 09:34:02 jenkins2-159 dockerd[30572]: time="2017-01-25T09:34:02.461168590-05:00" level=debug msg="Loaded container f907113187893f2c60a66140fb40ae51ba8e6708324e6a47df08967032dd522c"
Jan 25 09:34:02 jenkins2-159 dockerd[30572]: time="2017-01-25T09:34:02.483225112-05:00" level=warning msg="libcontainerd: client is out of sync, restore was called on a fully synced container (f907113187893f2c60a66140fb40ae51ba8e6708324e6a47df08967032dd522c)."
Jan 25 09:34:02 jenkins2-159 dockerd[30572]: time="2017-01-25T09:34:02.484597253-05:00" level=debug msg="libcontainerd: received past event &types.Event{Type:\"start-container\", Id:\"f907113187893f2c60a66140fb40ae51ba8e6708324e6a47df08967032dd522c\", Status:0x0, Pid:\"\", Timestamp:(*timestamp.Timestamp)(0xc4203c8080)}"
Jan 25 09:34:02 jenkins2-159 dockerd[30572]: time="2017-01-25T09:34:02.484688983-05:00" level=debug msg="libcontainerd: received past event &types.Event{Type:\"start-process\", Id:\"f907113187893f2c60a66140fb40ae51ba8e6708324e6a47df08967032dd522c\", Status:0x0, Pid:\"73f6407b9680b16061f2048ee6b566060451ea79c5020769544c552200209938\", Timestamp:(*timestamp.Timestamp)(0xc4203c81e0)}"
Jan 25 09:34:02 jenkins2-159 dockerd[30572]: time="2017-01-25T09:34:02.484738260-05:00" level=debug msg="libcontainerd: received past event &types.Event{Type:\"exit\", Id:\"f907113187893f2c60a66140fb40ae51ba8e6708324e6a47df08967032dd522c\", Status:0x7, Pid:\"73f6407b9680b16061f2048ee6b566060451ea79c5020769544c552200209938\", Timestamp:(*timestamp.Timestamp)(0xc4203c83e0)}"
Jan 25 09:34:02 jenkins2-159 dockerd[30572]: time="2017-01-25T09:34:02.484779067-05:00" level=debug msg="libcontainerd: received past event &types.Event{Type:\"start-process\", Id:\"f907113187893f2c60a66140fb40ae51ba8e6708324e6a47df08967032dd522c\", Status:0x0, Pid:\"c1572a29fc5a9b63a538bfdd2ab06f3973595722e64d256c141472f86550772b\", Timestamp:(*timestamp.Timestamp)(0xc4203c85b0)}"
Jan 25 09:34:02 jenkins2-159 dockerd[30572]: time="2017-01-25T09:34:02.484823355-05:00" level=debug msg="libcontainerd: received past event &types.Event{Type:\"exit\", Id:\"f907113187893f2c60a66140fb40ae51ba8e6708324e6a47df08967032dd522c\", Status:0x0, Pid:\"c1572a29fc5a9b63a538bfdd2ab06f3973595722e64d256c141472f86550772b\", Timestamp:(*timestamp.Timestamp)(0xc4203c8810)}"
Jan 25 09:34:02 jenkins2-159 dockerd[30572]: time="2017-01-25T09:34:02.484865136-05:00" level=debug msg="libcontainerd: received past event &types.Event{Type:\"start-process\", Id:\"f907113187893f2c60a66140fb40ae51ba8e6708324e6a47df08967032dd522c\", Status:0x0, Pid:\"74298e778837f767e6a5bcb1d1393f56cda98437104b166ef0bf54e960561e45\", Timestamp:(*timestamp.Timestamp)(0xc4203c8970)}"
Jan 25 09:34:02 jenkins2-159 dockerd[30572]: time="2017-01-25T09:34:02.484905280-05:00" level=debug msg="libcontainerd: received past event &types.Event{Type:\"exit\", Id:\"f907113187893f2c60a66140fb40ae51ba8e6708324e6a47df08967032dd522c\", Status:0x0, Pid:\"74298e778837f767e6a5bcb1d1393f56cda98437104b166ef0bf54e960561e45\", Timestamp:(*timestamp.Timestamp)(0xc4203c8a80)}"
Jan 25 09:34:02 jenkins2-159 dockerd[30572]: time="2017-01-25T09:34:02.484944278-05:00" level=debug msg="libcontainerd: received past event &types.Event{Type:\"exit\", Id:\"f907113187893f2c60a66140fb40ae51ba8e6708324e6a47df08967032dd522c\", Status:0x89, Pid:\"init\", Timestamp:(*timestamp.Timestamp)(0xc4203c8c00)}"
Jan 25 09:34:02 jenkins2-159 dockerd[30572]: time="2017-01-25T09:34:02.484979539-05:00" level=warning msg="libcontainerd: failed to retrieve container f907113187893f2c60a66140fb40ae51ba8e6708324e6a47df08967032dd522c state: rpc error: code = 2 desc = containerd: container not found"
```

@hmdhasani The fix for #30469 is now merged in master and v1.13.x branch. It will also be included in `v1.13.1-rc2`

**Comment by hmdhasani** [comment]
Excuse me for the delay.
I thinks it's fixed.

**Comment by thaJeztah** [comment]
Perfect, thanks!

**Comment by thaJeztah** [comment]
Let me lock the conversation on this issue; given that there's various reasons this error can occur; not all related to this one, I want to prevent this from collecting unrelated reports. If you have this issue on the current release, please open a new issue instead (and confirm there's no other open issue)
